Implement get_state, set_state, get_history, set_history in base class.

nabriis opened the pull request:
As described in #402 this PR seeks to implement get_state, set_state in the base class ´SamplerNew`. 

This is achieved by adding a class variable `_STATE_KEYS` that contain a set of keys representing the state variables the sampler will work with when calling the **already** implemented `get_state` and `set_state` methods. This reduces the amount of repeated code in each subclass.

Subclassing samplers must now _optionally_ update `_STATE_KEYS` if they expand on the set of variables the represent the state. For example the `ProposalBasedSampler` sampler can expand this by including the `scale` key, as this sampler maintains a `scale` attribute that is updated when running warmup sampling.

A testing framework is added get_state, set_state and checkpointing.

Additionally the `_HISTORY_KEYS` class variable is also added following the same design, but containing variable that representing the sampling history like the samples and acceptance criteria. This specific change introduces public methods get_history and set_history working similarly to how get_state and set_state already works before this PR.

Closes #402

The following related issue #379 is about refactoring the general behavior of get_state and set_state. This current implemented changes introduce 2 class variables that can relatively easily be replaced with another design once we agree on something else for #379.

amal-ghamdi commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -199,23 +199,3 @@ def tune(self, skip_len, update_count):
         self.scale = np.minimum(scale_temp, np.ones(self.dim))
         self._scale_temp = scale_temp
-
-    def get_state(self):
```
really great how much redundancy were removed (the definitions of get_state and set_state that were everywhere)!

amal-ghamdi commented on cuqi/experimental/mcmc/_pcn.py:
```diff
@@ -5,13 +5,15 @@
         self.scale = scale
         self.current_point = self.initial_point
-        self.current_loglike_eval = self._loglikelihood(self.current_point)
+        self.current_loglike_logd = self._loglikelihood(self.current_point)
```
suggest changing this variable to `current_likelihood_logd` (changing all occurrences of the variable) to be consistent with `current_target_logd`

nabriis replied:
Done

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -13,6 +13,27 @@ def progressbar(iterable, **kwargs):
         warnings.warn("Module mcmc: Progressbar not found. Install progressbar2 to get sampling progress.")
         return iterable
 
+class _UniqueList(list):
```
Is there a particular need for this data structure, instead of using Python set. for example:
```
set1 = {'key1'} # set elements are 'key1'
set1.add('key1') #  set elements are 'key1'
set1.add('key2') #  set elements are 'key1', 'key2'
set1.update({'key2', 'key3'})  #  set elements are 'key1', 'key2', 'key3'
```

nabriis replied:
done

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -45,8 +72,10 @@ def __init__(self, target: cuqi.density.Density, initial_point=None, callback=No
+        self.initial_point = initial_point
         
-        self._samples = [initial_point]
+        self._samples = [initial_point] # Remove?
```
Do we need to capture this in one of the issues we have or create a new issue (Consider whether to include initial_point in the initial state of self._samples?)

nabriis replied:
Done

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -104,6 +117,25 @@ def target(self, value):
         self._target = value
         self.validate_target()
 
+    @property
+    def current_point(self):
+        """ The current point of the sampler. """
+        return self._current_point
+    
+    @current_point.setter
+    def current_point(self, value):
+        """ Set the current point of the sampler. """
+        self._current_point = value
+
+    @property
+    def current_target_logd(self):
+        """ The current target value of the sampler. """
+        return self._current_target_logd
+    
+    @current_target_logd.setter
+    def current_target_logd(self, value):
+        """ Set the current target value of the sampler. """
+        self._current_target_logd = value
 
```
This comment probably not necessarily should be addressed in this PR. I think we need a way to make sure developers who create subclass of NewSampler do use these two properties. Here are two suggestion:
1- to enforce these as abstractmethods. To prevent a case when developer make a subclass of NewSampler and not consider using these properties at all. That is, it is part of the interface the developer needs to implement. This is I think is a tradeoff. It will lead to some redundancy, but it will make it clear that the developer need to take care of 'step' 'tune' and the two states: 'current_point', ' current_target_logd'.
2- as we did before, ask user to implement `_tune` and `_step` instead of `tune` and `step`. Then we have parent method `tune` and `step` to call `_tune` / `_step` expecting `current_point` and `target_logd` to be returned. and store them in state variable. So it is not the responsibility of  `_tune` / `_step` to store this state variable.

nabriis replied:
I agree there is something to consider here. I went with the following option:

Added a unit test to check if subclassing samplers utilize all state variables. See test_state_is_fully_updated_after_warmup_step.

We could still create an issue if you want to consider future, e.g. if users subclass within their ipynb or something.

We could add a "validate_sampler" method that can be run to check it works as expected.

amal-ghamdi replied:
really nice added tests!

I just created this issue #392 for a general user-defined sampler (I do not think we have one). If you think it makes sense, you can add the details you suggested in this issue.

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -204,6 +236,54 @@ def warmup(self, Nb, tune_freq=0.1) -> 'SamplerNew':
+    def get_state(self) -> dict:
+        """ Return the state of the sampler. """   
+        state = {
+            'metadata': {
```
it is good that you separated the metadata from what is required for state.

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -204,6 +236,54 @@ def warmup(self, Nb, tune_freq=0.1) -> 'SamplerNew':
+
+    def set_state(self, state: dict):
+        """ Set the state of the sampler. """
+        if state['metadata']['sampler_type'] != self.__class__.__name__:
```
nice :)

amal-ghamdi commented on tests/experimental/test_mcmc.py:
```diff
@@ -184,15 +185,54 @@ def test_CWMH_regression_warmup(target: cuqi.density.Density):
+        if cls not in skip_checkpoint  # use cls here, not name
+    ]
+
+    # Convert instances to their classes
```
```suggestion
    # Convert instances in checkpoint_targets to their classes
```

amal-ghamdi commented on tests/experimental/test_mcmc.py:
```diff
@@ -202,16 +242,54 @@ def test_checkpointing(sampler: cuqi.experimental.mcmc.SamplerNew):
     assert np.allclose(samples1, samples2), f"Samples1: {samples1}\nSamples2: {samples2}"
 
+
+@pytest.mark.parametrize("sampler", checkpoint_targets)
```
we can probably make these 2 tests for state keys and history keys general to all samplers not only those that are tested for checkpointing.

nabriis replied:
Done

amal-ghamdi reviewed (changes requested):
Thank you @nabriis for this great refactor! Looks much more robust than what we had. I added few comments.

jakobsj reviewed (commented):
Hi @nabriis, thank you for this, it is looking promising.  Like @amal-ghamdi I am wondering about the choice of data structure, whether a standard type like python set could be used. 

More generally, I was trying to work out how users should be able to interact with the state of samplers.  I think they should be able to easily view the current state of a sampler.  Load a state from disk, view it, perhaps modify, then set up a sampler with that state.  Perhaps even create a state from scratch (with easily seeing the expected entry types) and create a sampler.  I couldn't really see whether this was possible.  Is there a relevant demo outlining intended usage possibilities, to look at or can one be created?

It may be helpful to write out design/requirements to help see what the best data structure would be.

nabriis commented:
> Hi @nabriis, thank you for this, it is looking promising. Like @amal-ghamdi I am wondering about the choice of data structure, whether a standard type like python set could be used.
> 
> More generally, I was trying to work out how users should be able to interact with the state of samplers. I think they should be able to easily view the current state of a sampler. Load a state from disk, view it, perhaps modify, then set up a sampler with that state. Perhaps even create a state from scratch (with easily seeing the expected entry types) and create a sampler. I couldn't really see whether this was possible. Is there a relevant demo outlining intended usage possibilities, to look at or can one be created?
> 
> It may be helpful to write out design/requirements to help see what the best data structure would be.

I took your comment into account. The underlying datastructure to store the state internally in the sampler was updated to a set. The user is NOT exposed to this.

What the user is exposed to the what they can get from "get_state" and "set_state". These are dictionaries, and there is now more documentation on them.

The dictionary datastructure is useful because it is easy to extend what it can store (what keys it contains).

amal-ghamdi commented on cuqi/experimental/mcmc/_sampler.py:
```diff
@@ -238,7 +206,28 @@ def warmup(self, Nb, tune_freq=0.1) -> 'SamplerNew':
+            },
+            'state': {
+                'current_point': np.array([...]),
+                'current_target_logd': -123.45,
```
one thought here related to storing logd for all sampler. In some sense, logd is derived from `current_point`. If we want the sate to be minimal, maybe we restrict it to current point only without storing in logd values

nabriis replied:
You are right. I moved the logd out of the base sampler class. It now currently recides in ProposalBasedSampler. It can be refactored further in the future.

amal-ghamdi commented on tests/experimental/test_mcmc.py:
```diff
@@ -293,3 +302,56 @@ def test_history_keys(sampler: cuqi.experimental.mcmc.SamplerNew):
+
+# Dictionary to store keys that are not expected to be updated after warmup.
+# Likely due to not implemented feature in the sampler.
+state_exception_keys = {
```
(this is not directly relevant to this PR) it seems all these classes allow warmup but tuning is not implemented yet (thus scale is not updated). I suggest we add an item in issue #378 to say :

- [ ] For all the sampler classes in which tune is implemented as 'pass', do we need warning or error? to let the user know tuning is not happening?

amal-ghamdi reviewed (approved):
Thank you Nicolai, Looks great!

jakobsj commented:
Thanks! Please, for reference, add explanation as comment here on the PR on: 
- what you have done now in response to the latest comments and our discussions and decision around it, 
- what the latest commits seek to do, and 
- create an issue with explanation and the comments raised that have not been addressed by this PR and refer to it from here.

nabriis commented:
> Thanks! Please, for reference, add explanation as comment here on the PR on:
> 
> * what you have done now in response to the latest comments and our discussions and decision around it,
> * what the latest commits seek to do, and
> * create an issue with explanation and the comments raised that have not been addressed by this PR and refer to it from here.

Hello @jakobsj.

Thanks for requesting a clarification for future reference. I updated the issue #402 with a description of the sought change and updated this PR description with more details on the aim of this PR. Additionally, I included your comments into #379.

For reference this PR seeks to iterate on the current design by moving the get_state and set_state methods to the base class, not refactor the design in its entirety. Let us discuss that refactor in #379.

jakobsj reviewed (approved):
Many thanks for adding more explanation incl. scope, context and the capture of the review comments. This PR nicely addresses the issue of moving get/set state to the base class and managing state and history in a better way and provides a nice step forward toward at a later point addressing #379.
